src/main/java/org/assertj/core/api/RecursiveIterableComparisonAssert.java

```diff
@@ -0,0 +1,285 @@
+   * forces type compatibility of <strong>elements contained in iterables</strong> and {@link #withStrictTypeCheckingOnActualIterable() withStrictTypeCheckingOnActualIterable}
+   * forces type compatibility of the <strong>iterables themselves</strong>, not the elements contained within them.
+   * <p>
+   * <strong>Comparing ordered collections with unordered ones and vice/versa.</strong>
+   * By default, {@link #isEqualTo(Iterable) isEqualTo} takes element order into consideration.
+   * This means that a List and a Set can not be considered equals, even if they contain exactly the same elements.
+   * To ignore the element order when comparing iterables, call {@link #ignoringActualIterableOrder() ignoringActualIterableOrder}.
+   *
+   * <pre><code class='java'>public class Person {
+   *    String name;
+   *    boolean hasPhd;
+   * }
+   *
+   * public class Doctor {
+   *    String name;
+   *    boolean hasPhd;
+   * }
+   *
+   * Person sheldon = new People("Sheldon", true);
+   * Person leonard = new People("Leonard", true);
+   * Person howard = new People("Howard", false);
+   *
+   * Doctor drSheldon = new Doctor("Sheldon", false);
+   * Doctor drLeonard = new Doctor("Leonard", false);
+   *
+   *
+   * List<Person> peopleList = Arrays.asList(sheldon, leonard, howard);
+   * Set<Person> peopleLinkedSet = new LinkedHashSet<>(peopleList);
+   *
+   * // assertion succeeds because both iterables have same items in same order.
+   * assertThat(peopleList).usingRecursiveComparison()
+   *                       .isEqualTo(peopleLinkedSet);
+   *
+   *
+   * Set<Person> peopleHashSet = new HashSet<>(peopleList);
+   *
+   * // assertion fails even though iterables have same items, the order is different.
+   * assertThat(peopleList).usingRecursiveComparison()
+   *                       .allowingArrayTypeForExpected()
+   *                       .isEqualTo(peopleHashSet);
+   *
+   * // assertion succeeds because the order is ignored.
+   * assertThat(peopleList).usingRecursiveComparison()
+   *                       .ignoringActualIterableOrder()
+   *                       .isEqualTo(peopleHashSet);
+   *
+   * // assertion fails because iterables are restricted to be compatible by {@link #withStrictTypeCheckingOnActualIterable() withStrictTypeCheckingOnActualIterable}.
+   * assertThat(peopleList).usingRecursiveComparison()
+   *                       .withStrictTypeCheckingOnActualIterable()
+   *                       .isEqualTo(peopleLinkedSet);</code></pre>
+   *
+   * @param expected the object to compare {@code actual} to.
+   * @return {@code this} assertion object.
+   * @throws AssertionError if the actual and the given objects are not deeply equal property/field by property/field.
+   * @throws IntrospectionError if one property/field to compare can not be found.
+   */
+  @SuppressWarnings("unchecked")
+  public SELF isEqualTo(Iterable<?> expected) {
+    if((actual == null && expected == null) || (actual == expected))
```

Comment:
I believe `(actual == null && expected == null)` can be removed as we check `actual == null || expected == null` afterward

Reply:
The implementation should be like the start of `RecursiveComparisonAssert.isEqualTo`:
```
    // deals with both actual and expected being null
    if (actual == expected) return myself;
    if (expected == null) {
      // for the assertion to pass, actual must be null but this is not the case since actual != expected
      // => we fail expecting actual to be null
      objects.assertNull(info, actual);
    }
    // at this point expected is not null, which means actual must not be null for the assertion to pass
    objects.assertNotNull(info, actual);
```
Extract these common check to a protected method to avoid duplication.